src/vtquery.cpp

```diff
@@ -99,58 +55,136 @@ struct Worker : Nan::AsyncWorker
+            if (!radius_val->IsNumber())
+                return utils::CallbackError("'radius' must be a number", callback);
+
+            radius = radius_val->NumberValue();
+
+            if (radius < 0.0)
+                return utils::CallbackError("'radius' must be a positive number", callback);
+        }
+
+        if (options->Has(Nan::New("results").ToLocalChecked()))
+        {
+            v8::Local<v8::Value> results_val = options->Get(Nan::New("results").ToLocalChecked());
+            if (!results_val->IsNumber())
+                return utils::CallbackError("'results' must be a number", callback);
+
+            results = results_val->IntegerValue();
+
+            if (results < 0)
+                return utils::CallbackError("'results' must be a positive number", callback);
+        }
+
+        if (options->Has(Nan::New("layers").ToLocalChecked()))
+        {
+            v8::Local<v8::Value> layers_val = options->Get(Nan::New("layers").ToLocalChecked());
+            if (!layers_val->IsArray())
+                return utils::CallbackError("'layers' must be an array of strings", callback);
+
+            v8::Local<v8::Array> layers_arr = layers_val.As<v8::Array>();
+            unsigned num_layers = layers_arr->Length();
+
+            // only gather layers if there are some in the array
+            if (num_layers > 0)
+            {
+                custom_layers = true;
+                std::vector<std::string> layers;
+                layers.reserve(num_layers);
+                for (unsigned j=0; j < num_layers; ++j)
+                {
+                    v8::Local<v8::Value> layer_val = layers_arr->Get(j);
+                    if (!layer_val->IsString())
+                    {
+                        return utils::CallbackError("'layers' values must be strings", callback);
+                    }
+
+                    Nan::Utf8String layer_utf8_value(layer_val);
+                    int layer_str_len = layer_utf8_value.length();
+                    if (layer_str_len <= 0)
+                    {
+                        return utils::CallbackError("'layers' values must be non-empty strings", callback);
+                    }
+
+                    std::string layer(*layer_utf8_value, layer_str_len);
+                    layers.push_back(layer);
+                }
+            }
+        }
+
+        if (options->Has(Nan::New("geometry").ToLocalChecked()))
         {
-            return utils::CallbackError("option 'louder' must be a boolean",
-                                        callback);
+            custom_geometry = true;
+
+            v8::Local<v8::Value> geometry_val = options->Get(Nan::New("geometry").ToLocalChecked());
+            if (!geometry_val->IsString())
+                return utils::CallbackError("'geometry' option must be a string", callback);
+
+            Nan::Utf8String geometry_utf8_value(geometry_val);
```

Comment:
It is ideal to check for `geometry_utf8_value.length() <= 0` before trying to create a `std::string` since passing a null string may crash unpredictably (refs https://github.com/mapbox/node-fontnik/issues/124)

Reply:
Actually, while `std::string s(nullptr);` crashes, `std::string s(nullptr,0);` does not. So technically this usage is safe. However, it is still ideal to catch `geometry_utf8_value.length() <= 0` since a zero length string would be an invalid option value.

Reply:
@springmeyer I originally had this check, but since a couple lines down we are checking for string values == `point` `linestring` `polygon` and exiting if they don't, I figured this null string couldn't get too far. Happy to add it back in though!

Reply:
👍  on adding back.